cores/nRF5/Tone.cpp

```diff
@@ -34,35 +34,62 @@ Version Modified By Date     Comments
 0010	Arduino.org 16/07/27 Added Arduino Primo support
 *************************************************/
 
-
+#include "Arduino.h"
 #include "Tone.h"
 #include "WVariant.h"
 
-unsigned long int count_duration=0;
+volatile unsigned long int count_duration=0;
 volatile bool no_stop = false;
 uint8_t pin_sound=0;
+static uintptr_t _toneToken = 0x656e6f54; // 'T' 'o' 'n' 'e'
 
+// NOTE: Currently hard-coded to only use PWM2 ...
+//       These are the relvant hard-coded variables
+//       (plus the ISR PWM2_IRQHandler)
+static IRQn_Type      const _IntNo       = PWM2_IRQn;
+static NRF_PWM_Type * const _PWMInstance = NRF_PWM2;
+static HardwarePWM  * const _HwPWM       = HwPWMx[2];
 
 void tone(uint8_t pin, unsigned int frequency, unsigned long duration)
 {
+	static_assert(sizeof(unsigned long) == sizeof(uint32_t));
+	bool new_no_stop;
+	unsigned long int new_count_duration = (unsigned long int)-1L;
 	unsigned int time_per=0;
-	
-	if((frequency < 20) | (frequency > 25000)) return;
-	
+
+	// limit frequency to reasonable audible range	
+	if((frequency < 20) | (frequency > 25000)) {
+		LOG_LV1("TON", "frequency outside range [20..25000] -- ignoring");
+		return;
+	}
+
+	// set nostop to true to avoid race condition.
+	// Specifically, race between a tone finishing
+	// after checking for ownership (which releases ownership)
+	// No effect if a tone is not playing....
+	no_stop = true;
```

Comment:
<details><summary>Ok... click to expand</summary><P/>

Consider this function reaching new line # 116 (in the midst of configuring the peripheral with new values), and that the value of `count_duration == 1`.

Then the interrupt fires, so the ISR decrements the value, and it reaches zero.  The ISR will undo portions of the configuration, and then disable the peripheral.

When the ISR reverts control to this function, the configuration will be wrong, the peripheral will be disabled, etc....
</details>

Reply:
@henrygab I see, but it is not really needed since Arduino API requires user to call noTone() first which disable the IRQ before calling tone(). I think we can simplify this to just disable IRQ after taking the ownership of PWM.

https://www.arduino.cc/reference/en/language/functions/advanced-io/tone/

```
Notes and Warnings
If you want to play different pitches on multiple pins, you need to call noTone() on one pin before calling tone() on the next pin.
```

Reply:
no more race condition exists, due to removal of ISR in its entirety.